### ISSUE TITLE
3rd Party Deployment Issues

### ISSUE THREAD

**Issue by MattReimer** [Riverscapes/RiverscapesToolbar#6]
We have an issue where our plugin needs `Boto3` (and probably other things) but QGIS does not come with these things. 

The big problem here is that QGIS stock does not come with pip and even with OSGEO4W64 you still need to `easy_install pip` to get it. 

### Here's what I propose:

We repurpose the `https://github.com/Riverscapes/riverscapestools` repo to have all the common code and dependencies we need. It's already pip installable and it already requires `boto3`. We can 

#### Key Benefits

* This will handle needing more libraries in the future. We can add more dependencies to the pip repo.
* We can fail gracefully in the plugin to say "I'm sorry but you need to install the `riverscapestools` pip thing for this to work. Here are some instructions to do that: ..."
* We can put classes we use often like XSD validation, program traversal, s3 operations etc. in there and other people can write their own tools against it. 
* Instead of deprecating the command line we just include it with this repo so people can always work with riverscapes, even if they don't have QGIS

#### Possible downsides:

* Currently the QGIS plugin and the console have diverged. Bringing them back together could be work.
* @philipbaileynar pointed out that there is a danger of over-installing too many dependencies.
* Things like GDAL and Rasterio and others need to be compiled so simply listing them as dependencies won't work.

**Comment by jesselangdon** [comment]
It's an interesting idea. But I have a couple of clarifying questions.  

So would the riverscapes repo just have code that would load these dependices via pip?  Or would we actually create branches of the libraries and store them in the riverscapes repo (which would lock us into specific versions of each library)?

GDAL is included in OSGeo4W, but not Rasterio (which is why I've stayed away from using Rasterio/Fiona  /Shapely, even though they're nice libraries).  NetworkX is the main require I need at the moment in pyGNAT, but that could change once I build planned segmentation and catchment area delineation tools.

**Comment by MattReimer** [comment]
@jesselangdon what I'm proposing is that the `riverscapestools` repo would contain useful classes that we write but it would only contain a list of dependencies that pip would install for us:

[Check it out in the console pip repo](https://github.com/Riverscapes/riverscapestools/blob/master/setup.py#L13
)

so when I type

```
pip install git+https://github.com/Riverscapes/riverscapestools
```
it installs `boto3` and `argparese` for me if it's not already installed

I believe we can also lock in dependencies on specific versions of these 3rd party libraries so we're all using the same versions. 

You're right about `GDaL`, `rasterio`, `Fiona` etc. If we put these in as dependencies and the user hadn't compiled them the pip install of `riverscapestools` would fail. This means we need to be really careful about what we add to the list. 

I think you're right though that `networkx` is perfectly safe and could easily be added to the list.

**Comment by MattReimer** [comment]
@philipbaileynar This is a big refactor. I think we should wait and do it at the same time as the changes to the Riverscapes path structure. Otherwise we risk turning one big refactoring into 2.

**Comment by philipbaileynar** [comment]
@MattReimer  I completely agree. The priority right now is stabilizing RAVE and getting the existing functionality working:

* Download
* Opening projects
* symbology

For now, the tool will be used by technically competent analysts or by less technical users that have someone more experienced lead them through installation.

**Comment by MattReimer** [comment]
Cool. I'm gonna deprioritize this issue and we'll come back to it.
